pgx_lwt/src/pgx_lwt.ml

```diff
@@ -1,75 +1,65 @@
-open Lwt
+module S = S
 
 module Thread = struct
-  type 'a t = 'a Lwt.t
+  open Lwt
 
-  let return = return
-  let ( >>= ) = ( >>= )
-  let catch = catch
+  module Make (IO : S.IO) = struct
+    type 'a t = 'a Lwt.t
 
-  type sockaddr =
-    | Unix of string
-    | Inet of string * int
+    let return = return
+    let ( >>= ) = ( >>= )
+    let catch = catch
 
-  type in_channel = Lwt_io.input_channel
-  type out_channel = Lwt_io.output_channel
+    type sockaddr = IO.sockaddr =
+      | Unix of string
+      | Inet of string * int
 
-  let output_char = Lwt_io.write_char
-  let output_string = Lwt_io.write
+    type in_channel = IO.in_channel
+    type out_channel = IO.out_channel
 
-  let output_binary_int w n =
-    let chr = Char.chr in
-    output_char w (chr (n lsr 24))
-    >>= fun () ->
-    output_char w (chr ((n lsr 16) land 255))
-    >>= fun () ->
-    output_char w (chr ((n lsr 8) land 255))
-    >>= fun () -> output_char w (chr (n land 255))
-  ;;
+    let output_char = IO.output_char
+    let output_string = IO.output_string
 
-  let flush = Lwt_io.flush
-  let input_char = Lwt_io.read_char
-  let really_input = Lwt_io.read_into_exactly
+    let output_binary_int w n =
+      let chr = Char.chr in
+      output_char w (chr (n lsr 24))
+      >>= fun () ->
+      output_char w (chr ((n lsr 16) land 255))
+      >>= fun () ->
+      output_char w (chr ((n lsr 8) land 255))
+      >>= fun () -> output_char w (chr (n land 255))
+    ;;
 
-  let input_binary_int r =
-    let b = Bytes.create 4 in
-    Lwt_io.read_into_exactly r b 0 4
-    >|= fun () ->
-    let s = Bytes.to_string b in
-    let code = Char.code in
-    (code s.[0] lsl 24) lor (code s.[1] lsl 16) lor (code s.[2] lsl 8) lor code s.[3]
-  ;;
+    let flush = IO.flush
+    let input_char = IO.input_char
+    let really_input = IO.really_input
 
-  let close_in = Lwt_io.close
+    let input_binary_int r =
+      let b = Bytes.create 4 in
+      really_input r b 0 4
+      >|= fun () ->
+      let s = Bytes.to_string b in
+      let code = Char.code in
+      (code s.[0] lsl 24) lor (code s.[1] lsl 16) lor (code s.[2] lsl 8) lor code s.[3]
+    ;;
 
-  let open_connection sockaddr =
-    (match sockaddr with
-    | Unix path -> return (Unix.ADDR_UNIX path)
-    | Inet (hostname, port) ->
-      Lwt_unix.gethostbyname hostname
-      >|= fun { Lwt_unix.h_addr_list; _ } ->
-      let len = Array.length h_addr_list in
-      let i = Random.int len in
-      let addr = h_addr_list.(i) in
-      Unix.ADDR_INET (addr, port))
-    >>= Lwt_io.open_connection
-  ;;
+    let close_in = IO.close_in
+    let open_connection = IO.open_connection
+    let getlogin = IO.getlogin
+    let debug s = Logs_lwt.debug (fun m -> m "%s" s)
```

Comment:
It's sort of weird that we're assuming anyone using Lwt wants to use Logs, but I guess our logging just sucks in general.

Maybe one day we can make Pgx just require Logs in general..

Reply:
logs is the only logging option i know of that works with lwt so it felt like using it for the lwt backend was fine as its more likely to be pulled in by lwt users. If pgx itself ends up depending on logs then the `debug` implementation can be moved up to the core library as well.